### DISCUSSION TITLE
Show & Tell 48.5: ContinualLearningOrchestrator — Unified Lifelong Learning Pipeline

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#951] [Show and tell]
### The Stability-Plasticity Dilemma

The fundamental challenge: **stability** (retaining past knowledge) and **plasticity** (learning new information) are inherently in tension. Too much stability → the model cannot learn new tasks. Too much plasticity → catastrophic forgetting.

The orchestrator manages this balance dynamically across the model's lifetime.

### Strategy Selection by Scenario

```
Scenario Detection → Strategy Selection → Training Loop → Evaluation
     ↑                                                        ↓
     └──── Task Detector (48.4) ←── Metrics Feedback ────────┘
```

| Scenario | Best Primary Strategy | Best Hybrid |
|----------|---------------------|-------------|
| **Task-IL** | Progressive Nets (48.2) | ProgNets + EWC |
| **Domain-IL** | EWC/SI (48.1) | EWC + Replay |
| **Class-IL** | Experience Replay (48.3) | DER++ + EWC |

### Continual Learning Metrics

The orchestrator tracks a full **accuracy matrix** R[i][j] = accuracy on task j after training task i:

- **Average Accuracy**: Mean of final row R[T,:]
- **Backward Transfer (BWT)**: Average change in past task accuracy (negative = forgetting)
- **Forward Transfer (FWT)**: Accuracy on new tasks before training (positive = transfer)
- **Average Forgetting**: Max historical accuracy minus final accuracy per task

### Hybrid Strategy Composition

The orchestrator can combine strategies:

```python
# Example: EWC + Dark Experience Replay
loss = task_loss + lambda * ewc_penalty + alpha * dark_replay_loss
# With GEM gradient projection applied after backward pass
```

Multiple strategies often complement each other — EWC provides continuous regularization while replay provides episodic anchoring.

### Adaptive Lambda Scheduling

As tasks accumulate, the regularization strength may need adjustment:

```python
lambda_t = lambda_0 * (1 + decay_rate * forgetting_rate_t)
```

If forgetting increases → increase lambda. If plasticity is too low → decrease lambda.

### End-to-End Pipeline

1. **Data arrives** → TaskDetector monitors for distribution shift
2. **Drift detected** → Consolidate current task (Fisher, exemplars, column freeze)
3. **New task begins** → Select strategy based on scenario + resources
4. **Training** → Mixed batches (current + replay), regularized loss, projected gradients
5. **Evaluate** → Update accuracy matrix, compute CL metrics
6. **Adapt** → Adjust lambda, replay ratio, buffer composition based on metrics

Share your orchestration designs and benchmark results below!

Related: Issue #942 | Planning: #937
